server/src/main/java/org/elasticsearch/index/fielddata/ordinals/GlobalOrdinalsIndexFieldData.java

```diff
@@ -219,6 +221,25 @@ public SortedSetDocValues getOrdinalsValues() {
                         return values;
                     }
                     final TermsEnum[] atomicLookups = getOrLoadTermsEnums();
+                    /*
+                     * If we can manage it we try to use a wrapped singleton.
+                     * Lots of other code tries to unwrap the singleton with
+                     * DocValues.unwrapSingleton. It'll take a fast path if
+                     * it gets a singleton.
+                     *
+                     * We can manage a singleton if the total value count
+                     * fits in an `int` *and* the segment ords are a singleton.
+                     * We need the first one just because SortedDocValues
+                     * returns `int` instead of `long`
+                     */
+                    if (ordinalMap.getValueCount() < Integer.MAX_VALUE) {
+                        SortedDocValues singleton = DocValues.unwrapSingleton(values);
+                        if (singleton != null) {
+                            return DocValues.singleton(
+                                new SingletonGlobalOrdinalMapping(ordinalMap, singleton, atomicLookups, context.ord)
+                            );
+                        }
+                    }
```

Comment:
This feels a bit brittle. Basically, SingletonGlobalOrdinalMapping seems to be a very specialized class that can only function if its constructor parameters satisfy these very specific criteria that are neither documented nor enforced in the class itself. I wonder if we could make this less brittle if we made SingletonGlobalOrdinalMapping constructor private and moved the logic and comment above into a factory method that would return null in case the conditions are not satisfied.

Reply:
Yeah. The second test - that you can unwrap - is encoded in the ctor's signature. But the first bit - the value count - that totally should be guarded like you say.